Tests/NIOTests/SALChannelTests.swift

```diff
@@ -0,0 +1,236 @@
+            // change.
+            XCTAssertTrue(writableNotificationStepExpectation.compareAndExchange(expected: 3, desired: 4))
+            try self.assertWrite(expectedFD: .max,
+                                 expectedBytes: buffer,
+                                 return: .processed(2))
+
+            // And lastly, after having written everything, we'd expect to unregister for write
+            try self.assertReregister { selectable, eventSet in
+                XCTAssert(selectable as? Socket === channel.socket)
+                XCTAssertEqual([.read, .reset, .readEOF], eventSet)
+                return true
+            }
+
+            try self.assertParkedRightNow()
+        }) { () -> EventLoopFuture<Void> in
+            channel.setOption(ChannelOptions.writeSpin, value: 0).flatMap {
+                channel.setOption(ChannelOptions.writeBufferWaterMark, value: .init(low: 1, high: 1))
+            }.flatMap {
+                channel.pipeline.addHandler(DoWriteFromWritabilityChangedNotification(writableNotificationStepExpectation: writableNotificationStepExpectation))
+            }.flatMap {
+                // This write should cause a Channel writability change.
+                XCTAssertTrue(writableNotificationStepExpectation.compareAndExchange(expected: 0, desired: 1))
+                return channel.writeAndFlush(buffer)
+            }
+        }.teleportWait())
+    }
+
+    func testBasicRead() {
+        let localAddress = try! SocketAddress(ipAddress: "0.1.2.3", port: 4)
+        let serverAddress = try! SocketAddress(ipAddress: "9.8.7.6", port: 5)
+
+        final class SignalGroupOnRead: ChannelInboundHandler {
+            typealias InboundIn = ByteBuffer
+
+            private let group: DispatchGroup
+            private var numberOfCalls = 0
+
+            init(group: DispatchGroup) {
+                self.group = group
+            }
+
+            func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+                self.numberOfCalls += 1
+                XCTAssertEqual("hello",
+                               String(decoding: self.unwrapInboundIn(data).readableBytesView, as: Unicode.UTF8.self))
+                if self.numberOfCalls == 1 {
+                    self.group.leave()
+                }
+            }
+        }
+
+        var maybeChannel: SocketChannel? = nil
+        XCTAssertNoThrow(maybeChannel = try self.makeConnectedSocketChannel(localAddress: localAddress,
+                                                                            remoteAddress: serverAddress))
+        guard let channel = maybeChannel else {
+            XCTFail("couldn't construct channel")
+            return
+        }
+
+        var buffer = ByteBufferAllocator().buffer(capacity: 5)
+        buffer.writeString("hello")
+
+        let g = DispatchGroup()
+        g.enter()
+
+        // So we can exhaust it with 2 bytes.
```

Comment:
This comment seems unrelated to the test.

Reply:
weird, yes, thanks.